controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,1008 @@
+
+   TimerRegs *tmr = reinterpret_cast<TimerRegs *>(TIMER2_BASE);
+
+   // Set the frequency
+   tmr->reload = (CPU_FREQ / pwmFreqHz)-1;
+
+   // Configure channel 2 in PWM output mode
+   tmr->ccMode[0] = 0x6800;
+
+   tmr->ccEnable = 0x10;
+
+   // Start with 0% duty cycle
+   tmr->compare[1] = 0;
+
+   // Load the shadow registers
+   tmr->event = 1;
+
+   // Start the counter
+   tmr->ctrl[0] = 0x81;
+}
+
+// Set the PWM period.  Currently the value is 0 to 255 which is
+// a shame.  We should change this to a float, say with a value
+// of 0.0 to 1.0 to make better use of our resolution
+void HalApi::analogWrite(PwmPin pin, int value) {
+
+   // Convert the value to a float
+   float duty =  value * (1.0/255);
+
+   TimerRegs *tmr;
+   int chan;
+   switch( pin )
+   {
+      case PwmPin::BLOWER:
+         tmr = reinterpret_cast<TimerRegs *>(TIMER2_BASE);
+         chan = 1;
+         break;
+   }
+
+   tmr->compare[chan] = tmr->reload * duty;
+}
+
+/******************************************************************
+ * Serial port to GUI
+ * Chapter 38 of the reference manual defines the USART registers.
+ *****************************************************************/
+
+// This class is a generic circular buffer for byte sized data.
+// It could probably go in it's own header outside the HAL, we
+// would just need to add interrupt suspend/restore to the HAL
+// definition.
+//
+// Note that this class is used from both the main line of code
+// and the interrupt handlers, so it needs to be thread safe.
+// I'm disabling interrupts during the critical sections to
+// ensure that's the case.
+template <int N> class CircBuff
+{
+   volatile uint8_t buff[N];
+   volatile int head, tail;
+
+public:
+   CircBuff(){
+      head = tail = 0;
+   }
+
+   // Return number of bytes available in the buffer
+   int FullCt(){
+      bool p = IntSuspend();
+      int ct = head-tail;
+      IntRestore(p);
+      if( ct < 0 ) ct += N;
+      return ct;
+   }
+
+   // Return number of free spaces in the buffer
+   int FreeCt(){
+      return N-1-FullCt();
+   }
+
+   // Get the oldest byte from the buffer.
+   // Returns -1 if the buffer is empty
+   int Get()
+   {
+      int ret = -1;
+
+      bool p = IntSuspend();
+
+      if( head != tail )
+      {
+         ret = buff[tail++];
+         if( tail >= N ) tail = 0;
+      }
+
+      IntRestore( p );
+      return ret;
+   }
+
+   // Add a byte to the buffer
+   // Returnes true on success, false if the buffer is full
+   bool Put( uint8_t dat )
+   {
+      bool ret = false;
+
+      bool p = IntSuspend();
+
+      int h = head+1;
+      if( h >= N ) h = 0;
+
+      if( h != tail )
+      {
+         buff[head] = dat;
+         head = h;
+         ret = true;
+      }
+
+      IntRestore( p );
+
+      return ret;
+   }
+};
+
+class UART
+{
+   CircBuff<128> rxDat;
+   CircBuff<128> txDat;
+   UART_Regs *reg;
+public:
+   void Init( uint32_t baseAddr, int baud )
```

Comment:
This is somewhere that I think strong typing (i.e. moving the reinterpret_cast out to the constants that define the various UART_regs base addresses) would be really nice.

Reply:
I'll work on switching to strong typed pointers in the header if that's what it takes to get this PR done.  Can't work on it until tonight though, still trying to hold down a day job

Reply:
I don't block the review on it; it's easy enough to go back and fix it.  Mostly we just need Eugene's LGTM.

Reply:
OK, I moved all the interpret_cast code for the register blocks to the header files.  I think this is what you have been looking for.  I then went back and tested all the basic functionality to make sure nothing broke in the process.  Looks good.